src/set_theory/surreal.lean

```diff
@@ -269,28 +433,168 @@ instance : has_lt surreal   := ⟨lt⟩
+    exact lt_of_nsmul_pos_lt hm hxy },
+  { exfalso,
+    rwa [← int.neg_succ_not_pos m] }
+end
+
+lemma gmul_cancel_of_pos {m : ℤ} {x y : surreal} (hm : 0 < m) (hmxy : m • x  = m • y) : x = y :=
+begin
+    contrapose hmxy,
+    cases ne_iff_lt_or_gt.mp hmxy with hmxy' hmxy',
+    { apply ne_of_lt,
+      exact lt_of_gsmul_pos_lt hm hmxy' },
+    { apply ne_of_gt,
+      exact lt_of_gsmul_pos_lt hm hmxy' }
+end
+
+/-- The surreal number `half`. -/
+def half : surreal := ⟦⟨pgame.half, pgame.numeric_half⟩⟧
+
+/-- Powers of the surreal number `half`. -/
+def pow_half (n : ℕ) : surreal := ⟦⟨pgame.pow_half n, pgame.numeric_pow_half⟩⟧
+
+@[simp] lemma pow_half_zero : pow_half 0 = 1 := rfl
+
+@[simp] lemma pow_half_one : pow_half 1 = half := rfl
+
+@[simp] theorem add_half_self_eq_one : half + half = 1 :=
+quotient.sound pgame.add_half_self_equiv_one
+
+lemma double_pow_half_succ_eq_pow_half (n : ℕ) : 2 • pow_half n.succ = pow_half n :=
+begin
+  rw two_nsmul,
+  apply quotient.sound,
+  exact pgame.add_pow_half_succ_self_eq_pow_half,
+end
+
+lemma nsmul_pow_two_pow_half (n : ℕ) : 2 ^ n • pow_half n = 1 :=
+begin
+  induction n with n hn,
+  { simp only [nsmul_one, pow_half_zero, nat.cast_one, pow_zero] },
+  { rw [← hn, ← double_pow_half_succ_eq_pow_half n, smul_smul (2^n) 2 (pow_half n.succ),
+        mul_comm, pow_succ] }
+end
+
+lemma nsmul_pow_two_pow_half' (n k : ℕ) : 2 ^ n • pow_half (n + k) = pow_half k :=
+begin
+  induction k with k hk,
+  { simp only [add_zero, surreal.nsmul_pow_two_pow_half, nat.nat_zero_eq_zero, eq_self_iff_true,
+               surreal.pow_half_zero] },
+  { rw [← double_pow_half_succ_eq_pow_half (n + k), ← double_pow_half_succ_eq_pow_half k,
+        smul_algebra_smul_comm] at hk,
+    exact @gmul_cancel_of_pos 2 _ _ (by norm_num) hk }
+end
+
+lemma nsmul_int_pow_two_pow_half (m : ℤ) (n k : ℕ) :
+  (m * 2 ^ n) • pow_half (n + k) = m • pow_half k :=
+begin
+  rw mul_gsmul,
+  congr,
+  norm_cast,
+  exact nsmul_pow_two_pow_half' n k,
+end
+
+lemma dyadic_aux {m₁ m₂ : ℤ} {y₁ y₂ : ℕ} (h₂ : m₁ * (2 ^ y₁) = m₂ * (2 ^ y₂)) :
+  m₁ • pow_half y₂ = m₂ • pow_half y₁ :=
+begin
+  obtain h | h := le_or_lt y₁ y₂,
+  { obtain ⟨c, hc⟩ := le_iff_exists_add.mp h,
+    rw [hc, add_comm, pow_add, ← mul_assoc, mul_eq_mul_right_iff] at h₂,
+    cases h₂,
+    { rw [h₂, hc, add_comm, nsmul_int_pow_two_pow_half m₂ c y₁] },
+    { have := nat.one_le_pow y₁ 2 nat.succ_pos',
+      linarith } },
+  { obtain ⟨c, hc⟩ := le_iff_exists_add.mp (le_of_lt h),
+    rw [hc, add_comm, pow_add, ← mul_assoc, mul_eq_mul_right_iff] at h₂,
+    cases h₂,
+    { rw [← h₂, hc, add_comm, nsmul_int_pow_two_pow_half m₁ c y₂] },
+    { have := nat.one_le_pow y₂ 2 nat.succ_pos',
+      linarith } }
+end
+
+/-- The map `dyadic_map` sends ⟦⟨m, 2^n⟩⟧ to m • half ^ n. -/
+def dyadic_map : localization (submonoid.powers (2 : ℤ)) → surreal :=
+begin
+  apply quotient.lift,
+  swap,
+  { rintro ⟨m, n⟩,
+    exact m • pow_half (submonoid.log n) },
```

Comment:
It's usually better to define data outside of tactic mode:
```suggestion
def dyadic_map (x : localization (submonoid.powers (2 : ℤ))) : surreal :=
quotient.lift_on' x (λ x : ℤ × submonoid.powers 2, x.1 • pow_half (submonoid.log x.2)) $ begin
```

Reply:
Thanks! Will keep this in mind in the future.